README.md
@author Zaki-Zhou

@email : zakizhou94@gmail.com

# dba-tools工具集
### 安装部署
```shell
cd dba-tools
sh install.sh
source ~/.bashrc
```
### 配置文件说明

#### common/vars_conf.sh

可以修改的变量配置，变量详细介绍如下: 

- MYSQL_CONFIG_DIR : 存放MySQL配置文件的目录
- MYSQL_DIR : MySQL的bin目录
- MONGO_DIR : Mongo的bin目录
- PYTHON_DIR : Python包的目录
- MYSQL_USER: MySQL使用的用户名，最好有所有权限
- MONGO_USER： Mongo使用的用户名，最好有所有权限
- DDB_USER: 连接DDB的用户名
- DDB_PASSWORD: 连接DDB的密码


#### common/mysql_pwd_list

mysql可用的密码，可配置多个，每个密码占一行

#### common/mongo_pwd_list

mongo可用的密码，可配置多个，每个密码占一行

### 工具介绍

​	本工具包一共包含3个小工具，dbs/dba/go，dbs用于数据库日常的显示，dba用于数据库日常的操作，go用于数据库日常的连接，具体内容请往下看。

#### dbs

**使用方法:** dbs

**作用:** 快速显示一个机器上面的所有mysql、mongo、ddb实例和相关信息

**效果展示:** 更多参数可以通过 dbs --help进行获取，详细如下: 

```mysql
[dba@db-dg150-198 dba-tools]$ dbs --help
Usage: dbs use to get MySQL、DDB、Mongo server info
Options:
    -m | --more         Print details info.				   # 打印更多详细信息
    -d | --database     Print database info.			   # 打印数据库的信息
    -h | --help         Print the help page of scripts.    # 查看帮助文档
```

执行dbs获取MySQL、Mongo的基础信息，详细如下: 

```mysql
[dba@db-dg150-198 dba-tools]$ dbs
 IP: 127.0.0.1       Command: mysql      Date: 2018-05-11 12:51:23
+------+--------------------+-----------------------------+------------------+---------------------+--------------+
| Port | Vesion             | Defaults_File               | Datadir          | Master_Host_Port    | IO|SQL|Dealy |
+------+--------------------+-----------------------------+------------------+---------------------+--------------+
| 3306 | 5.5.23-rel25.3-log | /home/dba/mysqlnode/my1.cnf | /data1/nodedata1 |                     |              |
| 4306 | 5.7.21-log         | /home/dba/mysqlnode/my2.cnf | /data2/nodedata2 | 127.0.0.2:8306 | Yes|Yes|0    |
| 5306 | 5.7.21-log         | /home/dba/mysqlnode/my3.cnf | /data2/nodedata3 |                     |              |
| 6306 | 5.7.21-log         | /home/dba/mysqlnode/my4.cnf | /data3/nodedata4 | 127.0.0.2:8306 | Yes|Yes|0    |
| 8306 | 5.7.21-log         | /home/dba/mysqlnode/my5.cnf | /data2/nodedata5 |                     |              |
+------+--------------------+-----------------------------+------------------+---------------------+--------------+
 IP: 127.0.0.1       Command: mongo      Date: 2018-05-11 12:51:23
+-------+--------+-------------------------------------------------------+
| Port  | Server | Config_file                                           |
+-------+--------+-------------------------------------------------------+
| 20000 | mongod | /home/dba/sysbench-data/mongo/20000/mongod_20000.conf |
| 20001 | mongod | /home/dba/sysbench-data/mongo/20001/mongod_20001.conf |
| 32001 | mongod | /home/dba/config/shard.conf                           |
| 32003 | mongod | /home/dba/config/shard2.conf                          |
| 32004 | mongod | /home/dba/config/shard3.conf                          |
+-------+--------+-------------------------------------------------------+
```

#### go

**使用方法:** go port

**作用:** 通过端口快速连接一个mysql、mongo、ddb

**效果展示:** 连接mysql详细如下: 

```mysql
[dba@db-dg150-198 dba-tools]$ go 4306
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 38109
Server version: 5.7.21-log MySQL Community Server (GPL)

Copyright (c) 2000, 2018, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

(root@127.0.0.1:4306) [(none)]>
```

连接mongo详细如下: 

```mysql
[dba@db-dg150-198 dba-tools]$ go 32001
MongoDB shell version: 3.2.5
connecting to: 127.0.0.1:32001/admin
Server has startup warnings:
2018-04-08T10:38:29.308+0800 I CONTROL  [initandlisten]
2018-04-08T10:38:29.308+0800 I CONTROL  [initandlisten] ** WARNING: You are running on a NUMA machine.
2018-04-08T10:38:29.308+0800 I CONTROL  [initandlisten] **          We suggest launching mongod like this to avoid performance problems:
2018-04-08T10:38:29.308+0800 I CONTROL  [initandlisten] **              numactl --interleave=all mongod [other options]
2018-04-08T10:38:29.308+0800 I CONTROL  [initandlisten]
2018-04-08T10:38:29.308+0800 I CONTROL  [initandlisten] ** WARNING: soft rlimits too low. rlimits set to 65536 processes, 1000000 files. Number of processes should be at least 500000 : 0.5 times number of files.
2018-04-08T10:38:29.308+0800 I CONTROL  [initandlisten]
>
```

连接ddb详细如下: 

```mysql
[dba@db-39 ~]$ go 8888
User/password of administrator is given and user/password of normal user is omitted,
we will use user/password of administrator for normal user.
isql@dba>>
```

#### dba

**使用方法:** dba command port

**作用:** 用于管理数据库，将日常手动的操作简化

**效果展示:** 执行dba来查看帮助文档，详细如下: 

```mysql
[dba@db-dg150-198 dba-tools]$ dba
Usage: dba [Action] mysql_port or dba [Script] -h
Action:
    status		Check mysql status.(Support all)		# 查看mysql状态
    start		Start mysql by port.(Support all)		# 启动mysql实例
    stop		Stop mysql by port.(Support all)		# 关闭mysql实例
    restart		Restart mysql by port.(Support all)		# 重启mysql实例
    slave		Show slave status by port.(Support all) # 打印slave信息
    session		Show processlist by port.(Support all)  # 打印processlist信息
    sql			Show Running SQL by port.(Support all)  # 打印正在运行的SQL信息
    setro		Set global read_only = 1 by port.(Support all) # 执行Set global read_only=1
    setrw		Set global read_only = 0 by port.(Support all) # 执行Set global read_only=0
    opensqlsafe		Set global sql_safe_updates = 1 by port.(Support all) # 执行Set global sql_safe_updates = 1
    closesqlsafe	Set global sql_safe_updates = 0 by port.(Support all) # 执行Set global sql_safe_updates = 0
    mysql		Print orz -lazy -innodb_rows -innodb_data -mysql info  # 展示mysql实时信息
    mongo		Print mongostat info.								   # 展示mongo实时信息
    repair		Repair slave error.(Only support mysql5.5)	# 自动修复Slave的错误，只支持5.5
    killuser	Kill the User's MySQL session by port.(Support all) # 杀掉某用户的全部连接
    killall		Kill all MySQL session by port.(Support all)  # 杀掉除了系统连接外的所有连接
Script:
    admin		Managing mysql instances.				# status/start/stop/restart MySQL
    exe			Used to execute SQL.					# 可用于批量支持SQL
    orz			MySQL Real-time monitoring				# 监控mysql的脚本调用
```

**举个例子:** 

你想查看改机器上所有实例的binlog_format，执行`dba exe all 'show variables like "binlog_format"'`

```mysql
[dba@db-dg150-198 dba-tools]$ dba exe all 'show variables like "binlog_format"'
+---------------+-------+
| Variable_name | Value |
+---------------+-------+
| binlog_format | ROW   |
+---------------+-------+
Port: 3306 execute: show variables like "binlog_format"
+---------------+-------+
| Variable_name | Value |
+---------------+-------+
| binlog_format | ROW   |
+---------------+-------+
Port: 4306 execute: show variables like "binlog_format"
+---------------+-------+
| Variable_name | Value |
+---------------+-------+
| binlog_format | ROW   |
+---------------+-------+
Port: 5306 execute: show variables like "binlog_format"
+---------------+-------+
| Variable_name | Value |
+---------------+-------+
| binlog_format | ROW   |
+---------------+-------+
Port: 6306 execute: show variables like "binlog_format"
+---------------+-------+
| Variable_name | Value |
+---------------+-------+
| binlog_format | ROW   |
+---------------+-------+
Port: 8306 execute: show variables like "binlog_format"
```

### Release notes

[release notes](./doc/RELEASE_NOTES.md)


doc/RELEASE_NOTES.md
## Release notes
####  2018-05-11
- Functionality Added or Changed
  - 新增dbs显示mongo基础信息
  - 新增dba mysql 3306, 用于实时查看mysql状态值
  - 新增dba repair 3306, 快速修复Slave有问题的表, 目前只支持mysql5.5的版本
  - 新增dba status/stop/start/restart all操作, 可以一键重启所有实例
  - 新增dba killuser/killall 操作, 可以通过用户名判断快速kill mysql session
  - dbs显示mysql slave状态不为双Yes的时候显示红色
  - dbs加上log_slave_updates参数的显示
  - dbs -m 新增sync_binlog/innodb_flush_log_at_trx_commit的显示
- Bugs Fixed
  - 重大BUG: dba status/stop/start/restart 操作对端口不能精确匹配(已修复)
  - dbs显示mongo的优化
  - 危险操作加上了用户交互验证

####  2018-03-21
- Functionality Added or Changed
  - 新增go连接ddb功能
  - 新增go连接mongo功能
  - 新增日志模块
  - dba管理模块新增部分功能
- Bugs Fixed
  - dbs删除操作数据库的功能，迁移到dba管理工具下

#### 2018-02-08
- Functionality Added or Changed
  - 新增dba快捷方式，用于执行dba日常操作的工具
  - 新增功能一键打开或者关闭所有实例read-only
  - 新增启动和关闭实例脚本，可以通过端口进行判断
  - 新功能dbs -d 显示实例下的数据库
  - dbs添加变量sql_safe_updates的展示
- Bugs Fixed
  - dbs进行展示的时候，通过port进行排序
  - go脚本登陆时显示登陆的用户
  - 每个实例检测密码,解决同一个机器上数据库实例不同的问题
  - 没有实例的机器输出本机器没有实例
  - 兼容mysql5.7版本,表GLOBAL_VARIABLES迁移到了performance_schema

#### 2018-01-04
- Functionality Added or Changed
  - 密码验证功能拆分成独立功能
  - dbs -m 添加配置文件的显示
- Bugs Fixed
  - 合并Master_Host、Master_Port两个字段
  - 代码优化，所有字段变成动态
  - 缩减字段名字，解决显示问题
  - 修复dbs出现重复端口的bug

#### 2017-12-26
- Functionality Added or Changed
  - dbs脚本添加了功能：dbs m查看更多信息功能
  - go脚本添加了功能：优先使用端口进行登录
  - dbs增加了使用readlink获取准确的数据目录
- Bugs Fixed
  - 修复字段名称不规范问题
  - 修复dbs脚本部分查询Bug


common/basic_function.sh
# ################################################################
#!/bin/bash
# auter：Zaki
# 目的：基本功能函数脚本和基础变量
# 时间：2017年12月27日
# ################################################################

# 进入脚本所在工作目录
cd `dirname $0`

# 导入基础变量
PROJECT_NAME="dba-tools"
PROJECT_DIR=`pwd | awk -F "${PROJECT_NAME}" '{print $1}'`"${PROJECT_NAME}"
source "${PROJECT_DIR}/common/vars_conf.sh" 

# 基础工具
BIN_MYSQL_DIR="${MYSQL_DIR}/bin"
BIN_MONGO_DIR="${MONGO_DIR}/bin"
BIN_GREP="/bin/grep"
BIN_WHICH="/usr/bin/which"
BIN_PYTHON="${PYTHON_DIR}/bin/python"
PYTHON_LIB_DIR="${PYTHON_DIR}/lib"
# 本机实际ip
LOCAL_IP="127.0.0.1"
INSTANCE_IP=`/sbin/ifconfig -a|grep inet|grep -v 127.0.0.1|grep -v 192.168.122.1|grep -v inet6|awk '{print $2}'|tr -d "addr:"|sort -n | head -n 1`
# 时间
DATE_TIME=`date "+%Y-%m-%d %H:%M:%S"`
DATE_TODAY=`date "+%Y%m%d"`
DATE_TIMESTAMP=`date "+%Y%m%d_%H%M%S"`
# 其他变量
SCRIPETS_NAME=`echo $0 | awk -F '.' '{print $1}' | awk -F '/' '{print $NF}'`
SCRIPETS_PATH=`cd "$(dirname $0)"; pwd`
RANDOM_STRING=`date +%s%N | md5sum | head -c 5`
# 日志存储
LOG_DIR="${SCRIPETS_PATH}/log"
COMMON_LOG_FILE="${LOG_DIR}/${SCRIPETS_NAME}_${DATE_TODAY}.log"
EXECUTE_LOG_FILE="`pwd`/log/execute${DATE_TODAY}.log"

# ################################################################
# mysql mongo 环境变量检测
# ################################################################
MYSQL_PROFILE=0
MONGO_PROFILE=0
if [ ! -d "${BIN_MYSQL_DIR}" ];then
    BIN_MYSQL=`${BIN_WHICH} mysql --skip-alias 2>/dev/null`
    if [ $? -ne 0 -o ! -n "${BIN_MYSQL}" ];then
        MYSQL_PROFILE=1
    else
    	BIN_MYSQL_DIR=`dirname ${BIN_MYSQL}`
    	if [ $? -ne 0 -o ! -n "${BIN_MYSQL_DIR}" ];then
            MYSQL_PROFILE=1
    	fi
    fi
fi

if [ ! -d "${BIN_MONGO_DIR}" ];then
    BIN_MONGO=`${BIN_WHICH} mongo --skip-alias 2>/dev/null`
    if [ $? -ne 0 -o ! -n "${BIN_MONGO}" ];then
        MONGO_PROFILE=1
    else
    	BIN_MONGO_DIR=`dirname ${BIN_MONGO}`
    	if [ $? -ne 0 -o ! -n "${BIN_MONGO_DIR}" ];then
            MONGO_PROFILE=1
    	fi
    fi
fi


if [ ${MYSQL_PROFILE} -eq 1 -a ${MONGO_PROFILE} -eq 1 ];then
   echo -e "\033[31m[ERROR] Please add mysql or mongo environment variable in the file: ~/.bash_profile.\033[0m" 
   exit 1
fi

BIN_MYSQL="${BIN_MYSQL_DIR}/mysql"
BIN_MYSQLDUMP="${BIN_MYSQL_DIR}/mysqldump"
BIN_MONGO="${BIN_MONGO_DIR}/mongo"
BIN_MONGOSTAT="${BIN_MONGO_DIR}/mongostat"

# ################################################################
# 函数功能：打印字体输出
# 调用方法：checkInstance variables
# ################################################################
printWhite(){
     print_data_length=$1
     print_data=$2
     printf "| %-${print_data_length}s " "${print_data}"
}

printRed(){
     print_data_length=$1
     print_data=$2
     printf "| \033[31m%-${print_data_length}s\033[0m " "${print_data}"
}

printGreen(){
     print_data_length=$1
     print_data=$2
     printf "| \033[32m%-${print_data_length}s\033[0m " "${print_data}"
}

printYellow(){
     print_data_length=$1
     print_data=$2
     printf "| \033[33m%-${print_data_length}s\033[0m " "${print_data}"
}

echoRed(){
     ECHO_TEXT=$1
     echo -e "\033[31m${ECHO_TEXT}\033[0m"
}

echoGreen(){
     ECHO_TEXT=$1
     echo -e "\033[32m${ECHO_TEXT}\033[0m"
}

echoYellow(){
     ECHO_TEXT=$1
     echo -e "\033[33m${ECHO_TEXT}\033[0m"
}

# ################################################################
# 函数功能：输出框架
# 调用方法：framePrintf 4 5 6
# ################################################################
framePrintf(){
    while [ "$1" != "" ]
    do
	frameNum=$1
        printf "+"
        for ((framenum=1;framenum<=${frameNum};framenum++))
        do
             printf "-"
        done
	shift
    done
    if [ ! -n "$1" ];then
    	printf "+\n"
    fi
}

# ################################################################
# 函数功能：判断一个数组最长字符串的长度
# 调用方法：getArrayMaxLength "${name[*]}"  name是一个数组变量
# ################################################################
getArrayMaxLength(){
    array="$1"

    # 初始化最大长度
    max_length=0
    for var in ${array[*]}
    do
	string_length=${#var}
	if [ ${string_length} -gt ${max_length} ];then
	    max_length=${string_length}
	fi
    done
    echo ${max_length}
}

# ################################################################
# 函数功能：检测mysql密码的功能
# 调用方法：getArrayMaxLength socket全路径
# ################################################################
checkMysqlPasswd(){
    socket=$1
    password_file=$2
    # 文件位置
    if [ ! -n "${password_file}" ];then
    	basepath=$(cd `dirname $0`; pwd)
    	password_file="$basepath/common/mysql_pwd_list"
    fi
    while read password
    do
        user="root"
        test_row=`${BIN_MYSQL} -u${user} -p${password} --socket=${socket} -e "select 1" 2>/dev/null`
        if [ $? -eq 0 ];then
            echo ${password}
            return 0
        fi
    done < $password_file
    echo ""
}

# ################################################################
# 函数功能：检测mongo密码的功能
# 调用方法：checkMongoPasswd
# ################################################################
checkMongoPasswd(){
    mongo_port=$1
    password_file=$2
    # 文件位置
    if [ ! -n "${password_file}" ];then
    	basepath=$(cd `dirname $0`; pwd)
    	password_file="$basepath/common/mongo_pwd_list" 
    fi
    while read mongo_pwd
    do
	test_row=`${BIN_MONGO} -u${MONGO_USER} -p${mongo_pwd} --host ${INSTANCE_IP} --port ${mongo_port} admin --eval "1" 2>/dev/null`
	if [ $? -eq 0 ];then
            echo ${mongo_pwd}
            return 0
        fi
    done < $password_file
    echo ""
}

# ################################################################
# 函数功能：检测实例是否为空
# 调用方法：checkInstance "xxx" variables
# ################################################################
checkInstance(){
    print_log=$1
    check_variable=$2 
    if [ ! -n "${check_variable}" ];then
	echo -e "${print_log}"
	exit 1
    fi
}

checkFile(){
    print_log=$1
    check_variable=$2
    if [ ! -f "${check_variable}" ];then
	echoRed "${print_log}"
	exit 1
    fi
}

checkDir(){
    print_log=$1
    check_variable=$2
    if [ ! -d "${check_variable}" ];then
	echoRed "${print_log}"
	exit 1
    fi
}

# ################################################################
# 函数功能：操作crontab
# 调用方法：
# ################################################################
closeCrontabCheckSlave(){
     DATE_TIMESTAMP=`date "+%Y%m%d_%H%M%S"`
     TMP_FILE="/tmp/crontab${DATE_TIMESTAMP}"
     crontab -l > ${TMP_FILE}
     sed -i '/slave/s/^/#&/' ${TMP_FILE}
     crontab ${TMP_FILE}
     rm ${TMP_FILE}
}

openCrontabCheckSlave(){
     DATE_TIMESTAMP=`date "+%Y%m%d_%H%M%S"`
     TMP_FILE="/tmp/crontab${DATE_TIMESTAMP}"
     crontab -l > ${TMP_FILE}
     sed -i '/slave/s/^#//' ${TMP_FILE}
     crontab ${TMP_FILE}
     rm ${TMP_FILE}
}

###########################################################################
# 功能: 获取XML格式配置的某项值
# 使用示例：parsingXmlConfig "/home/ddb/DBClusterConf.xml" pid
# 输出: master-8888.pid
###########################################################################
parsingXmlConfig(){
    XML_CONFIG_FILE=$1
    XML_OPTIONS_NAME=$2
    echo `cat ${XML_CONFIG_FILE} | grep -w ${XML_OPTIONS_NAME} | sed -e "s/.*<${XML_OPTIONS_NAME}>\(.*\)<\/${XML_OPTIONS_NAME}>.*/\1/g"`
}

###########################################################################
# 功能: 打印日志
# 使用示例：printLog "printLog"
# 输出: 2018-01-11 11:23:03 printLog
###########################################################################
checkLogdir(){
   if [ ! -d "${LOG_DIR}" ];then
        mkdir "${LOG_DIR}"
   fi
}

printLog(){
   # 检查文件夹是否存在
   checkLogdir

   INFO_LOG_CONTENTS=$1
   echo -e "[INFO] ${DATE_TIME} ${INFO_LOG_CONTENTS}"
   echo -e "[INFO] ${DATE_TIME} ${INFO_LOG_CONTENTS}" >> ${COMMON_LOG_FILE}
}

printWarning(){
   # 检查文件夹是否存在
   checkLogdir

   WARNING_LOG_CONTENTS=$1
   echoYellow "[WARNING] ${DATE_TIME} ${WARNING_LOG_CONTENTS}"
   echo -e "[WARNING] ${DATE_TIME} ${WARNING_LOG_CONTENTS}" >> ${COMMON_LOG_FILE}
}

printError(){
   # 检查文件夹是否存在
   checkLogdir

   ERROR_LOG_CONTENTS=$1
   echoRed "[ERROR] ${DATE_TIME} ${ERROR_LOG_CONTENTS}"
   echoRed "[ERROR] ${DATE_TIME} Please check the log ${COMMON_LOG_FILE}."
   echo -e "[ERROR] ${DATE_TIME} ${ERROR_LOG_CONTENTS}" >> ${COMMON_LOG_FILE}
   echo -e "[ERROR] ${DATE_TIME} Please check the log ${COMMON_LOG_FILE}." >> ${COMMON_LOG_FILE}
}

###########################################################################
# 功能: 判断返回码是否有错误
# 使用示例:
# 返回：
###########################################################################
checkOk(){
    RETURN_CODE=$?
    if [ ${RETURN_CODE} -ne 0 ];then
	exit 1
    fi
}

checkReturnCode(){
     RETURN_CODE=$1
     if [ ${RETURN_CODE} -eq 0 ];then
        printLog "Success"
     else
        printError "Failed.Please check."
     fi
}

###########################################################################
# 功能: 换算单位,将多少b转换为相应的M/G/T
# 使用示例：
# 输出: 
###########################################################################
unitConversion(){
    CONVERSION=$1
    if [ -n "${CONVERSION}" ];then
        UNIT=0
        result=${CONVERSION}
        result_tmp=`expr ${CONVERSION} / 1024`
        while [ ${result_tmp} -ge 1 ]
        do
            result=${result_tmp}
            result_tmp=`expr ${result} / 1024`
            (( UNIT++ ))
        done
        case ${UNIT} in
            0)
	    result="${result}B"
            ;;
            1)
	    result="${result}K"
            ;;
            2)
	    result="${result}M"
            ;;
            3)
	    result="${result}G"
            ;;
            4)
	    result="${result}T"
            ;;
           # 其他单位报错
            *)
            exit 1
            ;;
        esac
    fi
}

###########################################################################
# 功能: 获取配置文件的绝对路径
# 使用示例：getTrueConfig ./my1.cnf 12345
# 返回参数:result
###########################################################################
getTrueConfig(){
    # 传入参数
    CONFIG_FILE_TMP=$1
    SERVER_PID=$2 
    
    # 判断传入参数是否为空
    if [ ! -n "${CONFIG_FILE_TMP}" -o ! -n "${SERVER_PID}" ];then
	return 1
    fi

    # 获取绝对路径
    if [[ "${CONFIG_FILE_TMP}" =~ "./" ]];then
	CONFIG_FILE_TMP=`echo ${CONFIG_FILE_TMP} | sed 's/^.\///g'`
	SERVER_START_DIR=`pwdx ${SERVER_PID} | awk -F ':' '{print $2}' | sed 's/^[ \t]*//g'`
	CONFIG_FILE=${SERVER_START_DIR}/${CONFIG_FILE_TMP}
    else
	CONFIG_FILE="${CONFIG_FILE_TMP}"
    fi
    
    # 返回
    result=${CONFIG_FILE}
    return 0
}

###########################################################################
# 功能: 解析json字符串
# 使用示例：resolveJson "string" "关键字"
# 返回参数: 关键字的值
###########################################################################
resolveJson(){
    JSON_STRING="$1"
    RESOLVE_NAME="$2"
    # 判断传入参数是否为空
    if [ ! -n "${JSON_STRING}" -o ! -n "${RESOLVE_NAME}" ];then
        return 1
    fi
    
    # 开始解析
    RESOLVE_VALUE=`echo "${JSON_STRING}" | awk -F "\"${RESOLVE_NAME}\" : " '{print $2}' | awk -F',' '{print $1}'`
    if [ $? -ne 0 ];then
	return 1
    fi

    # 返回
    result="${RESOLVE_VALUE}"
    return 0
}

###########################################################################
# 功能: 用于危险操作警告
# 使用示例：输入yes跳过
# 返回参数: 
###########################################################################
dangerWarning(){
    # 传入需要输出的文字
    DANGER_TEXT="$1"
    
    # 进行输入
    printf "\033[31m[Warning]%s. Are You Sure? [Y/n]\033[0m" "${DANGER_TEXT}"
    read -r -p "" input
    
    # 进行确认是否为yes
    case $input in
        [yY][eE][sS]|[yY])
                    return
                    ;;
    
        [nN][oO]|[nN])
                    exit 1
                    ;;
    
        *)
            exit 1
            ;;
    esac
}

# ################################################################
# 函数功能：通过端口获取mysql datadir
# 使用方法：getMysqlDatadir datadir
# ################################################################
getMysqlDatadir(){
    mysql_datadir=$1
    if [ -n "${mysql_datadir}" ];then
        mysql_real_datadir=`readlink ${mysql_datadir%?}`
        if [ ! -n "$mysql_real_datadir" ];then
            echo "${mysql_datadir}" | sed 's/[/]*$//g'
        else
            echo "${mysql_real_datadir}" | sed 's/[/]*$//g'
        fi
    fi
}

# ################################################################
# 函数功能：通过端口获取mysql socket
# 使用方法：getMysqlSocket 3306
# ################################################################
getMysqlSocket(){
    mysql_port=$1
    mysql_socket=`ps -ef | grep -w "port=${mysql_port}" | grep -v grep | awk '{if(match($0,".*socket=([^; ]*)",a))print a[1]}'`
    echo "${mysql_socket}"
}
